src/TheThingsNetwork.cpp

```diff
@@ -357,13 +357,18 @@ void TheThingsNetwork::clearReadBuffer()
   }
 }
 
-size_t TheThingsNetwork::readLine(char *buffer, size_t size)
+size_t TheThingsNetwork::readLine(char *buffer, size_t size, uint8_t attempts = 3) // Default timeout value 10s and is set in class initiator
 {
   size_t read = 0;
-  while (read == 0)
+  while (read == 0 && attempts>0)
   {
     read = modemStream->readBytesUntil('\n', buffer, size);
   }
+  if(attempts<=0)
+  { // If attempts is activated return 0 and set RN state marker
+    this->radioModuleInvalidState = true; // Inform the application about the radio module is not responsive.
```

Comment:
Although this is a way to signal, still it requires action from the application. I don't really know if application developers know how to resolve this.

The best way to resolve this is to reset the RN module through the reset pin. Would it be a good idea to configure (optionally) the reset pin, and engage a soft reset in the library at this point?

Reply:
I see your valid point. Now a good question is should we let the library handle the reset or give the decision to the application. If the library is resetting the RN module we will need to let the application know so the application can handle this, eg. by re-joining the network.
But in either way we must raise a flag to the application and abort the current task. 
If we reset the RN module in the library we can do a soft-reset by sending a sys_reset cmd or a hard-reset using the reset pin. In either way the library needs to rejoin the network and then re-execute the function. Should the module then retry for infinity or a limited tries?

One option could be to soft/hard reset the module every time this error occur, print a debug message with details and raise the flag. In this way we have raised awareness of a problem and tried to resolve it while giving power over the exception handling to the application.

readLine is used in:
- readResponse
- personalize
- join
- sendBytes
- waitForOk

Reply:
However maybe the simple solution is the best. On error exit with debug message and error flag. Then update documentation and make a good example on how to handle the error on application level.

Reply:
My wording was wrong; I meant hard reset, by using the reset pin. Sending the soft reset command is not going to work as the serial is broken already.

And you're right, doing a hard reset in the midst of a join or TX operation, and restoring the state (i.e. joining again) is not going to work either.

So I'm good with the signaling to the application. Please rename to `needsHardReset`, introduce a `resetHard()` that takes a pin number as parameter, and reset the `needsHardReset` after hard resetting. Also documentation would be helpful. Thanks!

Reply:
Okay. I will work on this and come back soon.